### PR TITLE
[lokiexporter] Add Resource Attributes as Loki Label  

### PR DESCRIPTION
**Pull request by eroteme**
**Description:** 
Logrecord attributes usually only give information at the log level (such a severity and request url, etc). 
Resource attributes usually add the context of what resource sent the log (such as container_name or cluster_name etc).

Currently the loki exporter is able to add logrecord attributes to loki labels. This PR adds Resource attributes also, which helps with filtering of loki data on both log attributes and resource attributes.

**Link to tracking Issue:** 
#3405

**Testing:** 
Added a few tests and a basic run of the exporter. I still have some questions about what happens if we have attributes with the same key on both log records and resources and which we would choose and how we deduplicate

**Documentation:** 
Updated README.md, added code comments for doc creation

### CODE REVIEW

File: exporter/lokiexporter/README.md
Comment (gregoryfranklin) [inline exporter/lokiexporter/README.md:62]: Apart from the misspelling of request, perhaps use a more typical example?  An access log containing http requests would use semantic conventions such as http.url, but this would not be suitable for a loki label.  http.host or http.status_code might make suitable loki labels though.
```suggestion
      http.status_code: "http_status_code" 
```

**Comment by gramidt** [comment]
Excellent work, @eroteme! From a quick pass everything looks good; however, I will allocate more time tomorrow to dive in and perform a deeper review. 

What are your thoughts on how you would like to handle duplicate labels?

**Comment by gregoryfranklin** [comment]
> What are your thoughts on how you would like to handle duplicate labels?

For duplicate labels, it might be worth considering the prometheus exporter as prior-art?
Loki labels are the same as OpenMetrics labels.  The spec defines how to convert OpenTelemetry attribute keys into OpenMetrics label keys, but doesn't mention merging keys from both Resource and LogRecord/Metric Attributes.
The prometheus exporter just overwrites the metric attributes with labels coming from the resource.  This seems reasonable to me.

Although, given you are explicitly listing which labels come from the Resource and which labels come from the LogRecord in the config, I don't think this is a big deal since you would have to explicitly configure a duplicate label situation.
Duplicate labels are therefore only an issue if the lokiexporter were later updated to support converting all attributes to labels instead of a specific list.

**Comment by gramidt** [comment]
> > What are your thoughts on how you would like to handle duplicate labels?
> 
> For duplicate labels, it might be worth considering the prometheus exporter as prior-art?
> Loki labels are the same as OpenMetrics labels. The spec defines how to convert OpenTelemetry attribute keys into OpenMetrics label keys, but doesn't mention merging keys from both Resource and LogRecord/Metric Attributes.
> The prometheus exporter just overwrites the metric attributes with labels coming from the resource. This seems reasonable to me.
> 
> Although, given you are explicitly listing which labels come from the Resource and which labels come from the LogRecord in the config, I don't think this is a big deal since you would have to explicitly configure a duplicate label situation.
> Duplicate labels are therefore only an issue if the lokiexporter were later updated to support converting all attributes to labels instead of a specific list.

Thank you so much for the prompt response! I personally like the idea of following the Prometheus exporter and overwriting the log attributes with the resource attributes. Let's try to do that in this PR.

**Comment by eroteme** [comment]
> 
> Thank you so much for the prompt response! I personally like the idea of following the Prometheus exporter and overwriting the log attributes with the resource attributes. Let's try to do that in this PR.

The prometheus model.labelset Merge function used
https://github.com/prometheus/common/blob/f45215cb5711443fffd211cab7671b268245c4a1/model/labelset.go#L118

just needed reversing in this case but I've added a comment to show that this is happening in the Merge

`mergedAttributes = logRecordAttributes.Merge(resourceAttributes)`

This should mean that the logrecord attributes are overwritten by the resource attributes

File: exporter/lokiexporter/config.go
Comment (gramidt) [inline exporter/lokiexporter/config.go:101]: Any change you had thoughts on combining the methods for LogRecordAttributes and ResourceAttributes? If not, no worries at all. Just curious if you had thought through it or not.

File: exporter/lokiexporter/exporter.go
Comment (gramidt) [inline exporter/lokiexporter/exporter.go:200]: Any change you had thoughts on combining these methods too? Seems to be a bit too much duplication if we can avoid it.

**Review by gramidt — CHANGES_REQUESTED** [review]

File: exporter/lokiexporter/config.go
Comment (eroteme) [reply exporter/lokiexporter/config.go:101]: I could add both loops to a single function and return 2 `map[string]model.LabelName`. We need to distinguish between the map of logRecordAttributes and ResourceAttributes so can't merge the two loop as far as I can tell

File: exporter/lokiexporter/exporter.go
Comment (eroteme) [reply exporter/lokiexporter/exporter.go:200]: I suppose I could add a sub struct for all labels (`map[string]ToLabels`) onto `lokiExporter` struct and have both logResourceAttribute labels and ResourceAttribute labels on that. This would mean I could merge the functions and would also work with the above duplicate code?

**Comment by eroteme** [comment]
I've tried to remove as much duplicated code by passing the attributes into the get attribute functions. I also noticed that the lokiExporter was storing the attribute labels when they where accessible  from the config anyway. Let me know if this is ok.

**Comment by gramidt** [comment]
@eroteme - This looks good. Could you rebase from main?

**Comment by gramidt** [comment]
@bogdandrutu @tigrannajaryan - Can you allow the pipeline to run? Since @eroteme is a first time contributor to this repository, GitHub workflows require a Maintainer to approve running the workflow.

**Comment by bogdandrutu** [comment]
@gramidt pipeline ran, but not sure what to do, can you approve if ready to merge?

**Review by gramidt — APPROVED** [review]

**Review by tigrannajaryan — APPROVED** [review]

**Comment by gramidt** [comment]
Thank you for your contribution, @eroteme!

**Pull request merged**
